File: crates/rocode-server/web/src/lib/multimodal.ts
export interface MultimodalAttachmentInfo {
  filename: string;
  mime: string;
}

export interface MultimodalPolicyView {
  voice: {
    duration_seconds: number;
    attach_audio: boolean;
    mime: string;
    language?: string | null;
  };
  max_input_bytes?: number | null;
  max_attachments_per_prompt?: number | null;
  allow_audio_input: boolean;
  allow_image_input: boolean;
  allow_file_input: boolean;
}

export interface MultimodalCapabilityView {
  provider_id: string;
  model_id: string;
  attachment: boolean;
  tool_call: boolean;
  reasoning: boolean;
  temperature: boolean;
  input: {
    text: boolean;
    audio: boolean;
    image: boolean;
    video: boolean;
    pdf: boolean;
  };
  output: {
    text: boolean;
    audio: boolean;
    image: boolean;
    video: boolean;
    pdf: boolean;
  };
}

export interface MultimodalPreflightResult {
  warnings: string[];
  unsupported_parts: string[];
  recommended_downgrade?: string | null;
  hard_block: boolean;
}

export interface MultimodalPreflightResponse {
  policy: MultimodalPolicyView;
  resolved_model?: string | null;
  capability?: MultimodalCapabilityView | null;
  result: MultimodalPreflightResult;
  warnings: string[];
}

export interface MultimodalCapabilitiesResponse {
  resolved_model?: string | null;
  capability?: MultimodalCapabilityView | null;
  warnings: string[];
}

export interface SessionPartRecord {
  type: string;
  url?: string;
  filename?: string;
  mime?: string;
}

export interface MultimodalPreflightRequest {
  model?: string;
  parts: unknown[];
  session_parts?: SessionPartRecord[];
}

export interface PersistedMultimodalExplain {
  attachment_count: number;
  kinds: string[];
  badges: string[];
  compact_label?: string | null;
  resolved_model?: string | null;
  warnings: string[];
  unsupported_parts: string[];
  recommended_downgrade?: string | null;
  hard_block: boolean;
  transport_replaced_parts: string[];
  transport_warnings: string[];
  attachments: MultimodalAttachmentInfo[];
}

export interface SessionMultimodalInsight extends PersistedMultimodalExplain {
  user_message_id: string;
}

export interface MultimodalInfoField {
  label?: string;
  value?: string;
}

export interface MultimodalExplainBlock {
  id: string;
  kind: "multimodal_info";
  role?: string;
  title: string;
  text: string;
  fields: MultimodalInfoField[];
}

export interface MultimodalHistoryBlock {
  id: string;
  kind: "message" | "multimodal_info" | "status";
  role?: string;
  title?: string;
  text: string;
  fields?: MultimodalInfoField[];
  tone?: "warning";
}

export function multimodalDisplayLabel(multimodal?: PersistedMultimodalExplain | null) {
  const compactLabel = multimodal?.compact_label?.trim();
  if (compactLabel) return compactLabel;
  const attachmentCount = multimodal?.attachment_count ?? 0;
  if (attachmentCount <= 0) return null;
  return attachmentCount === 1 ? "[1 attachment]" : `[${attachmentCount} attachments]`;
}

export function multimodalCombinedWarnings(multimodal?: PersistedMultimodalExplain | null) {
  const merged = [
    ...(multimodal?.warnings ?? []),
    ...(multimodal?.transport_warnings ?? []),
  ];
  return merged.filter((warning, index) => {
    const normalized = warning.trim();
    if (!normalized) return false;
    return merged.findIndex((candidate) => candidate.trim() === normalized) === index;
  });
}

export function buildMultimodalExplainBlocks(message: {
  id: string;
  role?: string;
  multimodal?: PersistedMultimodalExplain | null;
}): MultimodalExplainBlock[] {
  const multimodal = message.multimodal;
  if (!multimodal || multimodal.attachment_count <= 0) {
    return [];
  }

  const fields: MultimodalInfoField[] = [
    { label: "Attachments", value: String(multimodal.attachment_count) },
  ];
  if (multimodal.kinds.length > 0) {
    fields.push({ label: "Kinds", value: multimodal.kinds.join(", ") });
  }
  if (multimodal.resolved_model?.trim()) {
    fields.push({ label: "Target Model", value: multimodal.resolved_model.trim() });
  }
  if (multimodal.transport_replaced_parts.length > 0) {
    fields.push({
      label: "Transport Replaced",
      value: multimodal.transport_replaced_parts.join(", "),
    });
  }
  const attachmentNames = multimodal.attachments
    .map((file) => file.filename)
    .filter(Boolean)
    .join(", ");
  if (attachmentNames) {
    fields.push({ label: "Files", value: attachmentNames });
  }

  return [
    {
      id: `${message.id}:multimodal-info`,
      kind: "multimodal_info",
      role: message.role,
      title: "Multimodal Input",
      text: multimodalDisplayLabel(multimodal) ?? "[attachment input]",
      fields,
    },
  ];
}

export function buildMultimodalHistoryBlocks(message: {
  id: string;
  role?: string;
  multimodal?: PersistedMultimodalExplain | null;
}) {
  const blocks: MultimodalHistoryBlock[] = [];
  const fallbackText = multimodalDisplayLabel(message.multimodal);
  if (fallbackText) {
    blocks.push({
      id: `${message.id}:message`,
      kind: "message",
      role: message.role,
      text: fallbackText,
    });
  }

  blocks.push(...buildMultimodalExplainBlocks(message));

  const warnings = multimodalCombinedWarnings(message.multimodal);
  if (warnings.length > 0) {
    blocks.push({
      id: `${message.id}:multimodal-preflight`,
      kind: "status",
      tone: "warning",
      text: warnings.join(" "),
    });
  }

  return blocks;
}

File: crates/rocode-server/src/routes/multimodal.rs
use axum::{
    extract::{Query, State},
    routing::{get, post},
    Json, Router,
};
use rocode_multimodal::{
    MultimodalAuthority, MultimodalCapabilitiesResponse, MultimodalPolicyResponse,
    MultimodalPreflightRequest, MultimodalPreflightResponse, SessionPartAdapter,
};
use rocode_provider::bootstrap::{ProviderBootstrapState, ProviderModel};
use serde::Deserialize;
use std::sync::Arc;

use crate::server::bootstrap_config_from_config;
use crate::{ApiError, Result, ServerState};

pub(crate) fn multimodal_routes() -> Router<Arc<ServerState>> {
    Router::new()
        .route("/policy", get(get_multimodal_policy))
        .route("/capabilities", get(get_multimodal_capabilities))
        .route("/preflight", post(post_multimodal_preflight))
}

#[derive(Debug, Clone, Deserialize, Default)]
struct MultimodalCapabilitiesQuery {
    #[serde(default)]
    model: Option<String>,
}

async fn get_multimodal_policy(
    State(state): State<Arc<ServerState>>,
) -> Result<Json<MultimodalPolicyResponse>> {
    let config = state.config_store.config();
    let authority = MultimodalAuthority::from_config(&config);
    Ok(Json(MultimodalPolicyResponse {
        policy: authority.config_view(),
    }))
}

async fn get_multimodal_capabilities(
    State(state): State<Arc<ServerState>>,
    Query(query): Query<MultimodalCapabilitiesQuery>,
) -> Result<Json<MultimodalCapabilitiesResponse>> {
    let config = state.config_store.config();
    let authority = MultimodalAuthority::from_config(&config);
    let Some((provider_id, model)) = resolve_selected_model(&state, query.model.as_deref()).await?
    else {
        return Ok(Json(MultimodalCapabilitiesResponse {
            resolved_model: None,
            capability: None,
            warnings: vec!["No active model selected for multimodal capability lookup.".to_string()],
        }));
    };

    let capability = authority
        .capability_authority()
        .capability_view(provider_id.clone(), &model);
    Ok(Json(MultimodalCapabilitiesResponse {
        resolved_model: Some(format!("{}/{}", provider_id, model.id)),
        capability: Some(capability),
        warnings: Vec::new(),
    }))
}

async fn post_multimodal_preflight(
    State(state): State<Arc<ServerState>>,
    Json(payload): Json<MultimodalPreflightRequest>,
) -> Result<Json<MultimodalPreflightResponse>> {
    let config = state.config_store.config();
    let authority = MultimodalAuthority::from_config(&config);
    let policy = authority.config_view();
    let Some((provider_id, model)) =
        resolve_selected_model(&state, payload.model.as_deref()).await?
    else {
        return Ok(Json(MultimodalPreflightResponse {
            policy,
            resolved_model: None,
            capability: None,
            result: Default::default(),
            warnings: vec!["No active model selected for multimodal preflight.".to_string()],
        }));
    };

    let capability_authority = authority.capability_authority();
    let capability = capability_authority.capability_view(provider_id.clone(), &model);
    let effective_parts = if payload.session_parts.is_empty() {
        payload.parts
    } else {
        SessionPartAdapter::preflight_parts_from_session_parts(&payload.session_parts)
    };
    let result = capability_authority.preflight(&capability, &effective_parts);
    Ok(Json(MultimodalPreflightResponse {
        policy,
        resolved_model: Some(format!("{}/{}", provider_id, model.id)),
        capability: Some(capability),
        result,
        warnings: Vec::new(),
    }))
}

pub(crate) async fn resolve_selected_model(
    state: &Arc<ServerState>,
    explicit_model: Option<&str>,
) -> Result<Option<(String, ProviderModel)>> {
    let config = state.config_store.config();
    let model_ref = explicit_model
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
        .or_else(|| config.model.as_deref().map(str::trim).map(str::to_string))
        .filter(|value| !value.is_empty());

    let Some(model_ref) = model_ref else {
        return Ok(None);
    };

    let (provider_id, model_id) = if let Some((provider_id, model_id)) = model_ref.split_once('/') {
        (provider_id.to_string(), model_id.to_string())
    } else {
        let providers = state.providers.read().await;
        providers.parse_model_string(&model_ref).ok_or_else(|| {
            ApiError::BadRequest(format!(
                "unknown model `{}` for multimodal lookup",
                model_ref
            ))
        })?
    };

    let model = resolve_provider_model(state, &provider_id, &model_id).await?;
    Ok(Some((provider_id, model)))
}

pub(crate) async fn resolve_provider_model(
    state: &Arc<ServerState>,
    provider_id: &str,
    model_id: &str,
) -> Result<ProviderModel> {
    let config = state.config_store.config();
    let auth_store = state.auth_manager.list().await;
    let models_dev = state.catalog_authority.data().await;
    let bootstrap_config = bootstrap_config_from_config(&config);
    let bootstrap_state = ProviderBootstrapState::init(&models_dev, &bootstrap_config, &auth_store);

    let Some(provider) = bootstrap_state.providers.get(provider_id) else {
        return Err(ApiError::BadRequest(format!(
            "provider `{}` is not available for multimodal lookup",
            provider_id
        )));
    };
    let Some(model) = provider.models.get(model_id) else {
        return Err(ApiError::BadRequest(format!(
            "model `{}` was not found under provider `{}`",
            model_id, provider_id
        )));
    };

    Ok(model.clone())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ServerState;
    use rocode_config::{Config, ConfigStore, MultimodalConfig, ProviderConfig, VoiceConfig};
    use rocode_provider::{
        catalog::metadata_path_for_snapshot, ModelCatalogAuthority, ModelLimit, ModelModalities,
        ModelsData, ModelsDevInfo, ModelsProviderInfo,
    };
    use std::collections::HashMap;
    use tempfile::tempdir;

    fn test_state(config: Config, models_data: ModelsData) -> Arc<ServerState> {
        let temp = tempdir().expect("tempdir");
        let temp_path = temp.keep();
        let snapshot_path = temp_path.join("models.snapshot.json");
        std::fs::write(
            &snapshot_path,
            serde_json::to_vec(&models_data).expect("serialize snapshot"),
        )
        .expect("write snapshot");
        std::fs::write(
            metadata_path_for_snapshot(&snapshot_path),
            serde_json::to_vec(&rocode_provider::CatalogMetadata::default())
                .expect("serialize metadata"),
        )
        .expect("write metadata");

        let mut state = ServerState::new();
        state.config_store = Arc::new(ConfigStore::new(config));
        state.catalog_authority =
            Arc::new(ModelCatalogAuthority::with_snapshot_path(snapshot_path));
        Arc::new(state)
    }

    fn sample_models_data() -> ModelsData {
        let mut providers = HashMap::new();
        providers.insert(
            "openai".to_string(),
            ModelsProviderInfo {
                api: Some("https://api.openai.com/v1".to_string()),
                name: "OpenAI".to_string(),
                env: vec!["OPENAI_API_KEY".to_string()],
                id: "openai".to_string(),
                npm: Some("@ai-sdk/openai".to_string()),
                models: HashMap::from([(
                    "gpt-audio".to_string(),
                    ModelsDevInfo {
                        id: "gpt-audio".to_string(),
                        name: "GPT Audio".to_string(),
                        family: None,
                        release_date: Some("2026-01-01".to_string()),
                        attachment: true,
                        reasoning: false,
                        temperature: true,
                        tool_call: false,
                        interleaved: None,
                        cost: None,
                        limit: ModelLimit {
                            context: 128000,
                            input: None,
                            output: 4096,
                        },
                        modalities: Some(ModelModalities {
                            input: vec!["text".to_string(), "audio".to_string()],
                            output: vec!["text".to_string()],
                        }),
                        experimental: None,
                        status: Some("stable".to_string()),
                        options: HashMap::new(),
                        headers: None,
                        provider: None,
                        variants: None,
                    },
                )]),
            },
        );
        providers
    }

    #[tokio::test]
    async fn policy_route_returns_resolved_multimodal_policy() {
        let state = test_state(
            Config {
                multimodal: Some(MultimodalConfig {
                    voice: Some(VoiceConfig {
                        duration_seconds: Some(22),
                        attach_audio: Some(false),
                        ..Default::default()
                    }),
                    ..Default::default()
                }),
                ..Default::default()
            },
            sample_models_data(),
        );

        let Json(response) = get_multimodal_policy(State(state))
            .await
            .expect("policy route should succeed");
        assert_eq!(response.policy.voice.duration_seconds, 22);
        assert!(!response.policy.voice.attach_audio);
    }

    #[tokio::test]
    async fn preflight_route_uses_provider_ground_truth_audio_support() {
        let state = test_state(
            Config {
                model: Some("openai/gpt-audio".to_string()),
                provider: Some(HashMap::from([(
                    "openai".to_string(),
                    ProviderConfig {
                        name: Some("OpenAI".to_string()),
                        ..Default::default()
                    },
                )])),
                ..Default::default()
            },
            sample_models_data(),
        );

        let Json(response) = post_multimodal_preflight(
            State(state),
            Json(MultimodalPreflightRequest {
                model: None,
                parts: vec![rocode_multimodal::PreflightInputPart {
                    kind: Some(rocode_multimodal::ModalityKind::Audio),
                    mime: Some("audio/wav".to_string()),
                    byte_len: Some(512),
                    label: Some("voice.wav".to_string()),
                }],
                session_parts: Vec::new(),
            }),
        )
        .await
        .expect("preflight route should succeed");

        assert_eq!(response.resolved_model.as_deref(), Some("openai/gpt-audio"));
        assert!(response.capability.is_some());
        assert!(response.result.unsupported_parts.is_empty());
        assert!(!response.result.hard_block);
    }

    #[tokio::test]
    async fn preflight_route_accepts_session_parts_contract() {
        let state = test_state(
            Config {
                model: Some("openai/gpt-audio".to_string()),
                provider: Some(HashMap::from([(
                    "openai".to_string(),
                    ProviderConfig {
                        name: Some("OpenAI".to_string()),
                        ..Default::default()
                    },
                )])),
                ..Default::default()
            },
            sample_models_data(),
        );

        let Json(response) = post_multimodal_preflight(
            State(state),
            Json(MultimodalPreflightRequest {
                model: None,
                parts: Vec::new(),
                session_parts: vec![rocode_session::prompt::PartInput::File {
                    url: "data:audio/wav;base64,UklGRg==".to_string(),
                    filename: Some("voice.wav".to_string()),
                    mime: Some("audio/wav".to_string()),
                }],
            }),
        )
        .await
        .expect("preflight route should succeed");

        assert_eq!(response.resolved_model.as_deref(), Some("openai/gpt-audio"));
        assert!(response.result.warnings.is_empty());
        assert!(!response.result.hard_block);
    }
}
